torch/_inductor/spmd_prefetch.py

```diff
@@ -0,0 +1,128 @@
+# mypy: allow-untyped-defs
+# pyre-strict
+import typing
+from typing import (
+    Any,
+    Counter,
+    DefaultDict,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+from . import config, ir, comms, scheduler
+from collections import OrderedDict
+
+import torch
+
+from .dependencies import WeakDep
+from .utils import is_collective, is_wait, tuple_sorted
+
+
+    
+def reorder_all_gather(
+    snodes: List["scheduler.BaseSchedulerNode"],
+    all_gather_order: Optional[str] = "before",
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_order set to before: all_gather_i is reordered before wait_i-1
+    2. all_gather_order set to after: all_gather_i is reordered after wait_i-1
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
+    all_gather_list: List["scheduler.BaseSchedulerNode"] = []
+    node_type_list: List[int] = []
+
+    inverse_users, node_users = comms.compute_node_users(snodes)
+
+    for node in reversed(snodes):
+        node_type_list.append(get_node_type(node))
+
+    for idx, node in enumerate(reversed(snodes)):
+        node_type = node_type_list[idx]
+        if node_type in [2, 3]:
+            if node not in reorder_list and node not in all_gather_list:
+                reorder_list.append(node)
+        elif node_type == 1:
+            if all_gather_order == "after" and node_type_list[idx+1]==0 and len(all_gather_list) > 0:
+                reorder_list.extend(all_gather_list)
+                all_gather_list = []
+            reorder_list.append(node)
+            if all_gather_order == "before" and node_type_list[idx+1]==0 and len(all_gather_list) > 0: 
+                reorder_list.extend(all_gather_list)
+                all_gather_list = []
+        elif node_type == 0:
+            all_gather_list.append(node)
+            inverse_user = list(inverse_users[node])
+            if len(inverse_user) > 0:
+                all_gather_list.extend(inverse_user)
+        else:
+            raise ValueError("node type not supported")
+    if len(all_gather_list) > 0:
+        reorder_list.extend(all_gather_list)
+    reorder_list.reverse()
+    return reorder_list
+
+
+
+def reorder_reduce_scatter(
+    snodes: List["scheduler.BaseSchedulerNode"]
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder Reduce Scatter and Wait in the backward pass
+    reorder wait_i_rs before reduce_scatter_i+1 
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
+    wait_list: List["scheduler.BaseSchedulerNode"] = []
+    node_type_list: List[int] = []
+    
+    for node in snodes:
+        # ruisi: to get the convert_element_type after rs, can be removed after fixing the mpt bug.
```

Comment:
for these type of temporary comments, a common practice is to use `# TODO: ` or `# TODO(ruisizhang123)` when you would like to follow up later.